О данных

АНАЛИТИЧЕСКОЕ ОПИСАНИЕ ДАННЫХ:

Набор данных содержит информацию о продажах автомобилей BMW по всему миру за период с 2010 по 2024 год.
Данные включают характеристики автомобилей, их стоимость, пробег и объемы продаж в разных регионах.
ПЕРЕМЕННЫЕ В НАБОРЕ ДАННЫХ:
- Model: название автомобиля BMW (3-я серия, X5, i8 и т.д.)
- Year: год продажи (2010-2024)
- Region: географический рынок продаж (Азия, Европа, Северная Америка и др.)
- Color: цвет кузова автомобиля
- Fuel_Type: тип топливной системы (бензин, дизель, гибрид, электрический)
- Transmission: тип коробки передач (механическая, автоматическая)
- Engine_Size_L: объем двигателя в литрах
- Mileage_KM: пробег автомобиля в километрах
- Prise_USD: цена автомобиля в долларах США
- Sales_Volume: количество проданных единиц, объем продаж
- Sales_Classification: классификация продаж
- Clients: клиенты

In [2]:
#Загрузка библиотек
import numpy as np
import pandas as pd
from scipy import stats
import warnings
warnings.filterwarnings('ignore')

In [3]:
url = "https://gitlab.com/mariya20021801/data-analysis-in-python/-/raw/main/Project%201/Stepina.csv"
df = pd.read_csv(url)

# Предварительный анализ исходных данных

In [4]:
# На 1 этапе обязательно смотрим на данные (просто через df, через print(df), через df.head() или через df.tail()

In [5]:
df.head()

,Unnamed: 0,Model,Year,Region,Color,Fuel_Type,Transmission,Engine_Size_L,Mileage_KM,Price_USD,Sales_Volume,Sales_Classification,Clients
0,0,5 Series,2016.0,Asia,Red,Petrol,Manual,3.5,151748.0,NaN,8300.0,NaN,1
1,1,i8,NaN,North America,Red,Hybrid,NaN,1.6,NaN,79219.0,3428.0,Low,1
2,2,5 Series,NaN,North America,NaN,Petrol,Automatic,4.5,10991.0,NaN,6994.0,Low,1
3,3,X3,2024.0,Middle East,Blue,Petrol,Automatic,NaN,NaN,60971.0,NaN,Low,1
4,4,7 Series,2020.0,South America,Black,NaN,NaN,NaN,122131.0,49898.0,NaN,NaN,1


In [6]:
 df.tail()

,Unnamed: 0,Model,Year,Region,Color,Fuel_Type,Transmission,Engine_Size_L,Mileage_KM,Price_USD,Sales_Volume,Sales_Classification,Clients
49995,49995,i3,2014.0,Asia,NaN,NaN,Manual,4.6,NaN,NaN,8182.0,High,1
49996,49996,i3,2023.0,Middle East,Silver,NaN,NaN,4.2,147396.0,48714.0,NaN,High,1
49997,49997,NaN,NaN,Middle East,Red,Petrol,Automatic,4.5,NaN,46126.0,NaN,High,1
49998,49998,i3,2020.0,Asia,White,NaN,Automatic,3.8,NaN,NaN,9486.0,High,1
49999,49999,X1,2020.0,NaN,Blue,NaN,NaN,3.3,171003.0,NaN,1764.0,Low,1


In [7]:
# я хочу удалить первый и последний столбец, потому что на сайте Kaggle нет интерпритации, что это за столбец и что он обозначает, а первый потому что это дублирует индекс,который итак есть. 

In [8]:
df1 = df.drop(['Unnamed: 0', 'Clients'], axis=1)
df1.head()


,Model,Year,Region,Color,Fuel_Type,Transmission,Engine_Size_L,Mileage_KM,Price_USD,Sales_Volume,Sales_Classification
0,5 Series,2016.0,Asia,Red,Petrol,Manual,3.5,151748.0,NaN,8300.0,NaN
1,i8,NaN,North America,Red,Hybrid,NaN,1.6,NaN,79219.0,3428.0,Low
2,5 Series,NaN,North America,NaN,Petrol,Automatic,4.5,10991.0,NaN,6994.0,Low
3,X3,2024.0,Middle East,Blue,Petrol,Automatic,NaN,NaN,60971.0,NaN,Low
4,7 Series,2020.0,South America,Black,NaN,NaN,NaN,122131.0,49898.0,NaN,NaN


In [9]:
# предварительный просмотр первых 5 строк уже позволяет заметить наличие NaN

In [10]:
# просмотрим количество наблюдений в столбцах и типы данных
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 11 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Model                 31992 non-null  object 
 1   Year                  31955 non-null  float64
 2   Region                31886 non-null  object 
 3   Color                 31999 non-null  object 
 4   Fuel_Type             31964 non-null  object 
 5   Transmission          32164 non-null  object 
 6   Engine_Size_L         32103 non-null  float64
 7   Mileage_KM            32105 non-null  float64
 8   Price_USD             31933 non-null  float64
 9   Sales_Volume          31933 non-null  float64
 10  Sales_Classification  32138 non-null  object 
dtypes: float64(5), object(6)
memory usage: 4.2+ MB


Вывод: ОБЩАЯ ИНФОРМАЦИЯ:
- Размер данных: 50,000 записей, 11 столбцов
- Типы данных: 5 числовых (float64), 6 категориальных (object)
- Память: 4.2+ MB - оптимальный размер для анализа.
Так же кол-во данных разное, и это подтверждает наличие пропусков. И у показателя год стоит тип данных object, а должен int.

In [11]:
# Таблица с пропусками
пропуски_таблица = pd.DataFrame({
    'Столбец': df1.columns,
    'Пропущено_значений': df1.isnull().sum(),
    'Всего_значений': len(df1),
    'Процент_пропусков': (df1.isnull().sum() / len(df)) * 100
})

print(пропуски_таблица.round(2))

                                   Столбец  Пропущено_значений  \
Model                                Model               18008   
Year                                  Year               18045   
Region                              Region               18114   
Color                                Color               18001   
Fuel_Type                        Fuel_Type               18036   
Transmission                  Transmission               17836   
Engine_Size_L                Engine_Size_L               17897   
Mileage_KM                      Mileage_KM               17895   
Price_USD                        Price_USD               18067   
Sales_Volume                  Sales_Volume               18067   
Sales_Classification  Sales_Classification               17862   

                      Всего_значений  Процент_пропусков  
Model                          50000              36.02  
Year                           50000              36.09  
Region                         50

Вывод: 1. МАССОВЫЕ ПРОПУСКИ: Все столбцы имеют от ~36% до 38% пропущенных значений
- Model: 36.02% пропусков
- Year: 36.09% пропусков  
- Region: 36.23% пропусков
- Color 36.00%
- Fuel_Type 36.07%
- Engine_Size_L 35.79%
- Mileage_KM 35.79%  
- Price_USD: 36.13% пропусков
- Sales_Volume: 36.13% пропусков

2. НАИМЕНЕЕ ПРОБЛЕМНЫЕ (но все равно плохо):
   - Transmission: 35.67% пропусков
   - Sales_Classification: 35.72% пропусков

In [12]:
# далее я проведу заполнение пропусков, и преобразую Year в int.
# Преобразую Year в int
df1['Year'] = df1['Year'].fillna(df1['Year'].median()).astype(int)
#df1['Year'] = df1['Year'].astype(int)


In [13]:
# Создаю копию данных для очистки
df_clean = df1.copy()
# Выявление количественные и качественные признаки
numeric_cols = df_clean.select_dtypes(include=['int64', 'float64']).columns
categorical_cols = df_clean.select_dtypes(include=['object']).columns
# Заполняю количественные признаки средним (так было у вас в призентации)
df_clean[numeric_cols] = df_clean[numeric_cols].fillna(df_clean[numeric_cols].mean())
# Заполняю ка
df_clean[['Model', 'Region', 'Color', 'Fuel_Type', 'Transmission', 'Sales_Classification']] = df_clean[['Model', 'Region', 'Color', 'Fuel_Type', 'Transmission', 'Sales_Classification']].fillna(df_clean.mode().iloc[0])
# Удаляем строки с пропусками в качественных признаках
#df_clean = df_clean.dropna(subset=categorical_cols).reset_index(drop=True)

In [14]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 11 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Model                 50000 non-null  object 
 1   Year                  50000 non-null  int64  
 2   Region                50000 non-null  object 
 3   Color                 50000 non-null  object 
 4   Fuel_Type             50000 non-null  object 
 5   Transmission          50000 non-null  object 
 6   Engine_Size_L         50000 non-null  float64
 7   Mileage_KM            50000 non-null  float64
 8   Price_USD             50000 non-null  float64
 9   Sales_Volume          50000 non-null  float64
 10  Sales_Classification  50000 non-null  object 
dtypes: float64(4), int64(1), object(6)
memory usage: 4.2+ MB


Вывод: я заполнила пропуски количественных переменных - средним, а качественные переменные удалила, НО осталось очень мало значений. 
Осталось 3468 значений, я подумала, что это мало, тк нам нужно 10к, и я решила заполнить качественные переменные модой.
И тогда у меня получилось 50к значений, как и в первоначальном датасете.

In [15]:
# Проверяю, что пропусков больше нет
propyski = df_clean.isnull().sum().sum()
print(f"Осталось пропусков: {propyski}")

# Проверяю типы данных
print("\nТИПЫ ДАННЫХ ПОСЛЕ ОЧИСТКИ:")
print(df_clean.dtypes)


Осталось пропусков: 0

ТИПЫ ДАННЫХ ПОСЛЕ ОЧИСТКИ:
Model                    object
Year                      int64
Region                   object
Color                    object
Fuel_Type                object
Transmission             object
Engine_Size_L           float64
Mileage_KM              float64
Price_USD               float64
Sales_Volume            float64
Sales_Classification     object
dtype: object


# Проверяем наличие дубликатов

In [16]:
#Проверка на наличие дубликатов (выводятся только индексы одно из дубликатов, один остается в датафрейме)
duplicates = df_clean.duplicated()
duplicate_indexes = duplicates[duplicates].index.tolist()
print(f"Индексы строк-дубликатов: {duplicate_indexes}")
print(f"Количество строк-дубликатов: {len(duplicate_indexes)}")

Индексы строк-дубликатов: [947, 1202, 5962, 8041, 9496, 12662, 12893, 13038, 13300, 13547, 13771, 15117, 16316, 17343, 17413, 18042, 18642, 18733, 19293, 19969, 20787, 21039, 21167, 21753, 22582, 23008, 23890, 23972, 24457, 25076, 25306, 25436, 25870, 25965, 26075, 26368, 26919, 27133, 27297, 27715, 27876, 28898, 28977, 29350, 29580, 29862, 29985, 30853, 31209, 32465, 32707, 32724, 32922, 33407, 33521, 33530, 33555, 34544, 34712, 35137, 36432, 36530, 36651, 37465, 38289, 38574, 38885, 39526, 40031, 40403, 41726, 42015, 42200, 42445, 42643, 42653, 42779, 43169, 43702, 43914, 43950, 44242, 45132, 45336, 45368, 45462, 45967, 46432, 47084, 47180, 47223, 47418, 47699, 48742, 49280, 49659, 49931, 49986]
Количество строк-дубликатов: 98


In [17]:
# итого 98 дубликатов
#Так можно посмотреть все одинаковые строки
all_duplicates = df_clean[df_clean.duplicated(keep=False)]
print('Все дубли:', len(all_duplicates), 'шт.') #один из вариантов принта
# print(f"Все дублирующиеся строки ({len(all_duplicates)} шт.):")
all_duplicates.sort_index()

Все дубли: 165 шт.


,Model,Year,Region,Color,Fuel_Type,Transmission,Engine_Size_L,Mileage_KM,Price_USD,Sales_Volume,Sales_Classification
72,7 Series,2017,Asia,Red,Hybrid,Manual,4.50000,100590.857125,75105.375912,5090.692857,Low
100,7 Series,2017,Asia,Red,Hybrid,Manual,3.24532,100590.857125,75105.375912,5090.692857,Low
308,i8,2017,Asia,Red,Hybrid,Automatic,3.24532,100590.857125,75105.375912,5090.692857,Low
947,7 Series,2017,Asia,Red,Hybrid,Manual,3.24532,100590.857125,75105.375912,5090.692857,Low
1027,M5,2017,Asia,Grey,Diesel,Manual,3.24532,100590.857125,75105.375912,5090.692857,Low
...,...,...,...,...,...,...,...,...,...,...,...
48742,7 Series,2017,Middle East,Red,Diesel,Automatic,3.24532,100590.857125,75105.375912,5090.692857,Low
49280,3 Series,2017,Asia,Red,Electric,Automatic,3.24532,100590.857125,75105.375912,5090.692857,Low
49659,7 Series,2018,Asia,Red,Hybrid,Automatic,3.24532,100590.857125,75105.375912,5090.692857,Low
49931,7 Series,2012,Europe,Silver,Electric,Manual,3.24532,100590.857125,75105.375912,5090.692857,Low


In [18]:
#Удаляем строки с дублирующими индексами
df2 = df_clean.drop(duplicate_indexes).reset_index(drop=True)
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49902 entries, 0 to 49901
Data columns (total 11 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Model                 49902 non-null  object 
 1   Year                  49902 non-null  int64  
 2   Region                49902 non-null  object 
 3   Color                 49902 non-null  object 
 4   Fuel_Type             49902 non-null  object 
 5   Transmission          49902 non-null  object 
 6   Engine_Size_L         49902 non-null  float64
 7   Mileage_KM            49902 non-null  float64
 8   Price_USD             49902 non-null  float64
 9   Sales_Volume          49902 non-null  float64
 10  Sales_Classification  49902 non-null  object 
dtypes: float64(4), int64(1), object(6)
memory usage: 4.2+ MB


In [19]:
#Проверка на столбцы с 1 уникальным значением
df2.columns[df2.nunique().eq(1)]

Index([], dtype='object')

In [20]:
df2['Region'].unique() #можно просмотреть сами уникальные значение в этих столбцах

array(['Asia', 'North America', 'Middle East', 'South America', 'Europe',
       'Africa'], dtype=object)

Все столбцы содержат как минимум 2 разных значения. Данные разнообразны - нет константных столбцов
Не нужно удалять "бесполезные" столбцы на этом этапе очистки

In [21]:
df2.describe().round(2)

,Year,Engine_Size_L,Mileage_KM,Price_USD,Sales_Volume
count,49902.00,49902.00,49902.00,49902.00,49902.00
mean,2017.01,3.25,100590.86,75105.38,5090.69
std,3.45,0.81,46425.97,20811.31,2280.57
min,2010.00,1.50,29.00,30001.00,100.00
25%,2015.00,2.90,78513.75,65297.25,4044.00
50%,2017.00,3.25,100590.86,75105.38,5090.69
75%,2019.00,3.60,123399.75,85019.75,6172.75
max,2024.00,5.00,199991.00,119996.00,9999.00


In [22]:
# меня смутила цифра 9999.00 в Sales_Volume, возможно искусственное ограничение диапазона

In [23]:
df_9999 = df2.loc[df2['Sales_Volume'] == 9999.00]
df_9999.describe().round(2)

,Year,Engine_Size_L,Mileage_KM,Price_USD,Sales_Volume
count,2.00,2.00,2.00,2.00,2.0
mean,2020.50,3.65,76326.43,75105.38,9999.0
std,4.95,1.63,34315.08,0.00,0.0
min,2017.00,2.50,52062.00,75105.38,9999.0
25%,2018.75,3.08,64194.21,75105.38,9999.0
50%,2020.50,3.65,76326.43,75105.38,9999.0
75%,2022.25,4.22,88458.64,75105.38,9999.0
max,2024.00,4.80,100590.86,75105.38,9999.0


In [24]:
df_9999['Year'].unique()

array([2017, 2024])

In [25]:
# микровывод: что самый большой объем продаж приходится на период с 2017 по 2024 год. 
# Но гипотеза о том, что число закодировано, остается актуально и это держу у себя в голове. 

In [28]:
# при первоначальном анализе, не заметила старнных значений. Чуть подробнее пройдемся по показателям.
#YEAR (2010-2024):
#count    49902.00  - Анализируем 49,902 записи
#mean     2017.01   - Средний год продажи: 2017
#std         3.45   - Стандартное отклонение: 3.45 года
#min      2010.00   - Самая ранняя продажа: 2010 год
#25%      2015.00   - 25% продаж до 2015 года
#50%      2017.00   - Медиана: 2017 год (половина продаж до/после)
#75%      2019.00   - 75% продаж до 2019 года
#max      2024.00   - Самая поздняя продажа: 2024 год
#ENGINE_SIZE_L (1.5-5.0 литров):
#count    49902.00  - Все записи имеют значение объема двигателя
#mean         3.25   - Средний объем: 3.25 литра
#std          0.81   - Стандартное отклонение: 0.81 литра
#min          1.50   - Минимальный объем: 1.5L(малолитражки BMW)
#25%          2.90   - 25% автомобилей до 2.9L
#50%          3.25   - Медиана: 3.25L (типичный объем)
#75%          3.60   - 75% автомобилей до 3.6L
#max          5.00   - Максимальный объем: 5.0L (M-серия, спортивные)
#MILEAGE_KM (29 - 199,991 км):
#count    49902.00  - Все записи имеют значение пробега
#mean    100590.86  - Средний пробег: 100,591 км
#std      46425.97  - Большое стандартное отклонение
#min         29.00  - Минимальный пробег: 29 км
#25%     78513.75   - 25% автомобилей с пробегом до 78,514 км
##50%    100590.86   - Медиана: 100,591 км (СОВПАДАЕТ С СРЕДНИМ!)
#75%    123399.75   - 75% автомобилей с пробегом до 123,400 км
#max    199991.00   - Максимальный пробег: 199,991 км
#PRICE_USD ($30,001 - $119,996):
#count    49902.00  - Все записи имеют цену
##mean     75105.38  - Средняя цена: $75,105
#std      20811.31  - Стандартное отклонение: $20,811
#min      30001.00  - Минимальная цена: $30,001
#25%      65297.25  - 25% автомобилей дешевле $65,297
##50%      75105.38  - Медиана: $75,105
#75%      85019.75  - 75% автомобилей дешевле $85,020
#max     119996.00  - Максимальная цена: $119,996
#SALES_VOLUME (100 - 9,999 единиц):
#count    49902.00  - Все записи имеют объем продаж
##mean      5090.69  - Средний объем продаж: 5,091 единиц
#std       2280.57  - Стандартное отклонение: 2,281 единиц
#min        100.00  - Минимальный объем: 100 единиц
#25%       4044.00  - 25% моделей продано до 4,044 единиц
#50%       5090.69  - Медиана: 5,091 единиц
#75%       6172.75  - 75% моделей продано до 6,173 единиц
#max       9999.00  - Максимальный объем: 9,999 единиц

In [29]:
# Проверка года на корректность диапазона
print(f"Минимальный год: {df2['Year'].min()}")
print(f"Максимальный год: {df['Year'].max()}")

Минимальный год: 2010
Максимальный год: 2024.0


In [30]:
# Проверка цены на отрицательные значения
negative_price = (df2['Price_USD'] < 0).sum()
print(f"\nКоличество записей с отрицательной ценой: {negative_price}")


Количество записей с отрицательной ценой: 0


In [31]:
# Проверка пробега на отрицательные значения
negative_mileage = (df2['Mileage_KM'] < 0).sum()
print(f"Количество записей с отрицательным пробегом: {negative_mileage}")
# Проверка объема двигателя на корректность
print(f"\nДиапазон объема двигателя: {df2['Engine_Size_L'].min()} - {df2['Engine_Size_L'].max()} L")

Количество записей с отрицательным пробегом: 0

Диапазон объема двигателя: 1.5 - 5.0 L


Вывод: итого странные значения присутствуют, а именно максимальные значения в объеме продаж. Слишком аккуратные границы, скорее всего искуственное ограничение. 

# Вывод количества уникальных значений в каждом столбце

In [32]:
print(df2.nunique())

Model                      11
Year                       15
Region                      6
Color                       6
Fuel_Type                   4
Transmission                2
Engine_Size_L              37
Mileage_KM              29720
Price_USD               26891
Sales_Volume             9501
Sales_Classification        2
dtype: int64


In [33]:
#вывод: по категориальным переменным 
#Model (11 уникальных моделей):
 #  - Ограниченный модельный ряд BMW (11 моделей)
  # - Соответствует реальному портфелю BMW
   #- Позволяет анализировать популярность конкретных моделей
#Region (6 регионов):
 #  - Глобальное присутствие BMW в 6 регионах мира
  # - Типичное распределение: Европа, Азия, Северная Америка и т.д.
   #- Возможность сравнительного анализа рынков
#Color (6 цветов):
  # - Стандартная цветовая палитра BMW
  # - Обычно: черный, белый, серый, синий, красный, серебристый
   #- Анализ цветовых предпочтений по регионам/моделям
#Fuel_Type (4 типа топлива):
 #  - Бензин, дизель, гибрид, электрический
  # - Отражает эволюцию BMW к электромобилям
  # - Анализ трендов экологичных технологий
#Transmission (2 типа):
 #  - Автоматическая и механическая КПП
  # - Ожидаемое распределение (преобладание автомата)
   #- Анализ потребительских предпочтений
#Sales_Classification (2 категории):**
 #  - Вероятно: "Высокий спрос"/"Низкий спрос" или "Премиум"/"Стандарт
    

In [34]:
# вывод по кол-ым переменным
# Engine_Size_L (37 уникальных значений):
#    - Детальная градация объемов двигателей (например: 1.5, 1.6, 1.8...5.0)
#    - Точное моделирование технических характеристик
#    - Анализ зависимости цены от объема двигателя

# Mileage_KM (29,720 уникальных значений из 49,902 записей):
#    - ОЧЕНЬ ВЫСОКАЯ УНИКАЛЬНОСТЬ (почти 60% всех значений уникальны!)

# Price_USD (26,891 уникальных значений из 49,902 записей):
#    - ВЫСОКАЯ УНИКАЛЬНОСТЬ (~54%)
#    - Каждая цена почти уникальна

# Sales_Volume (9,501 уникальных значений):
#    - Широкий диапазон объемов продаж
#    - От 100 до 9,999 единиц (как ранее выявлено)

In [35]:
for columns in df2:
  print(df2[columns].unique())

['5 Series' 'i8' 'X3' '7 Series' 'M5' '3 Series' 'X5' 'X1' 'M3' 'i3' 'X6']
[2016 2017 2024 2020 2022 2019 2012 2014 2013 2011 2023 2015 2010 2021
 2018]
['Asia' 'North America' 'Middle East' 'South America' 'Europe' 'Africa']
['Red' 'Blue' 'Black' 'Silver' 'White' 'Grey']
['Petrol' 'Hybrid' 'Diesel' 'Electric']
['Manual' 'Automatic']
[3.5        1.6        4.5        3.24531975 1.7        2.4
 1.9        3.8        2.6        3.1        4.8        3.3
 3.2        4.1        2.         3.         2.1        4.3
 4.         3.4        3.7        3.9        1.5        4.2
 4.9        2.5        1.8        2.7        2.3        2.8
 2.2        4.7        4.6        4.4        2.9        3.6
 5.        ]
[151748.         100590.85712506  10991.         ...  56672.
 147396.         171003.        ]
[ 75105.37591207  79219.          60971.         ...  73173.
 115046.          48714.        ]
[8300. 3428. 6994. ... 6241. 2834. 6811.]
['Low' 'High']


In [36]:
# Результаты очистки данных, 49902 это кол-во наблюдений после очистки данных. А именно удаление дубликатов и пропуски я заполнила показателями.

# результаты квартильных групп 

In [37]:
q = df2.select_dtypes(include=['number']).quantile([0.1, 0.25, 0.5, 0.75, 0.95])
print(q)

        Year  Engine_Size_L     Mileage_KM      Price_USD  Sales_Volume
0.10  2012.0        2.00000   31147.300000   44233.100000   1695.000000
0.25  2015.0        2.90000   78513.750000   65297.250000   4044.000000
0.50  2017.0        3.24532  100590.857125   75105.375912   5090.692857
0.75  2019.0        3.60000  123399.750000   85019.750000   6172.750000
0.95  2023.0        4.70000  184832.750000  113185.850000   9233.000000



**Year (2012-2023):**

• 10-й процентиль: 2012 год - только 10% продаж было до 2012 года
• 25-й процентиль (Q1): 2015 год - 25% продаж пришлось на 2010-2015 гг.
• Медиана (Q2): 2017 год - половина всех продаж BMW совершена до 2017 года
• 75-й процентиль (Q3): 2019 год - 75% продаж было до 2019 года
• 95-й процентиль: 2023 год - 95% продаж совершено до 2023 года
ВЫВОДЫ:
1. Основной объем продаж BMW: 2015-2019 годы
2. Пик продаж пришелся на 2017-2019 годы
3. Данные охватывают активный период продаж (10 лет)



**Engine_Size_L (2.0L - 4.7L):**

• 10%: 2.0L - малолитражные модели BMW
• Q1: 2.9L - компактные и городские модели
• Median: 3.25L - СРЕДНИЙ ОБЪЕМ типичного BMW
• Q3: 3.6L - мощные седаны и кроссоверы
• 95%: 4.7L - спортивные и M-модели

ВЫВОДЫ:
1. Типичный BMW имеет объем 3.0-3.6L
2. 50% автомобилей в диапазоне 2.9-3.6L
3. Лишь 5% автомобилей имеют объем >4.7L



**Mileage_KM (31,147 - 184,833 км):**

• 10%: 31,147 км - почти новые автомобили (менее 3 лет)
• Q1: 78,514 км - автомобили с умеренным пробегом
• Median: 100,591 км - СРЕДНИЙ ПРОБЕГ типичного BMW
• Q3: 123,400 км - автомобили с высоким пробегом
• 95%: 184,833 км - автомобили с очень высоким пробегом

ВЫВОДЫ:
1. Половина BMW имеет пробег менее 100,000 км
2. Межквартильный размах: 78,514 - 123,400 км (диапазон типичных значений)
3. Лишь 5% автомобилей имеют пробег >184,833 км



**Price_USD ($44,233 - $113,186):**

• 10%: $44,233 - бюджетный сегмент BMW
• Q1: $65,297 - начало среднего ценового сегмента
• Median: $75,105 - СРЕДНЯЯ ЦЕНА типичного BMW
• Q3: $85,020 - премиальный сегмент
• 95%: $113,186 - люксовые и спортивные модели

ВЫВОДЫ:
1. Типичная цена BMW: $65,000-$85,000
2. 50% автомобилей стоят $65,297-$85,020
3. Разрыв между Q1 и Q3: ~$20,000 



**Sales_Volume (1,695 - 9,233 единиц):**

• 10%: 1,695 ед. - модели с низким спросом
• Q1: 4,044 ед. - стабильные, но не массовые модели
• Median: 5,091 ед. - СРЕДНИЙ ОБЪЕМ ПРОДАЖ
• Q3: 6,173 ед. - популярные модели
• 95%: 9,233 ед. - бестселлеры

ВЫВОДЫ:
1. Типичный объем продаж: 4,000-6,000 единиц
2. 50% моделей продается 4,044-6,173 единиц
3. Только 5% моделей продается >9,233 единиц


### Типичная машина (50-й процентиль):
   • Год: 2017
   • Объем двигателя: 3.25L
   • Пробег: 100,591 км
   • Цена: $75,105
   • Продажи: 5,091 единиц
### Какие можно увидеть зависимости: 
   • Более новые автомобили → выше цена
   • Больший объем двигателя → выше цена
   • Меньший пробег → выше цена
   • Популярность не всегда коррелирует с ценой

# группировка

In [38]:
df_gr = df2.groupby('Region')
round(df_gr.describe().T,2) #Т - транспонирование для удобства визуализации

Region                  Africa       Asia     Europe  Middle East  \
Year          count    5300.00   23467.00    5301.00      5366.00   
              mean     2016.95    2017.00    2017.04      2017.07   
              std         3.45       3.45       3.46         3.44   
              min      2010.00    2010.00    2010.00      2010.00   
              25%      2015.00    2015.00    2015.00      2015.00   
              50%      2017.00    2017.00    2017.00      2017.00   
              75%      2019.00    2019.00    2019.00      2019.00   
              max      2024.00    2024.00    2024.00      2024.00   
Engine_Size_L count    5300.00   23467.00    5301.00      5366.00   
              mean        3.26       3.24       3.25         3.25   
              std         0.82       0.81       0.81         0.81   
              min         1.50       1.50       1.50         1.50   
              25%         2.80       2.90       2.90         2.90   
              50%         3.25       3.25       3.25         3.25   
              75%         3.70       3.60       3.70         3.70   
              max         5.00       5.00       5.00         5.00   
Mileage_KM    count    5300.00   23467.00    5301.00      5366.00   
              mean   101488.31  100252.71  100583.86    100432.76   
              std     46712.04   46530.99   46713.53     45493.14   
              min        48.00      29.00     178.00        58.00   
              25%     80615.25   77616.50   78076.00     80248.25   
              50%    100590.86  100590.86  100590.86    100590.86   
              75%    124732.50  123518.50  123712.00    121841.50   
              max    199933.00  199991.00  199954.00    199941.00   
Price_USD     count    5300.00   23467.00    5301.00      5366.00   
              mean    74883.84   75058.53   75403.66     75178.05   
              std     20960.26   20771.05   20959.36     20973.44   
              min     30158.00   30001.00   30084.00     30008.00   
              25%     64939.25   65255.50   65967.00     65223.00   
              50%     75105.38   75105.38   75105.38     75105.38   
              75%     84952.00   84678.00   85820.00     85600.50   
              max    119988.00  119996.00  119981.00    119971.00   
Sales_Volume  count    5300.00   23467.00    5301.00      5366.00   
              mean     5096.30    5085.00    5108.82      5087.22   
              std      2304.14    2275.71    2294.18      2272.90   
              min       109.00     100.00     103.00       102.00   
              25%      3944.00    4030.00    4041.00      4062.25   
              50%      5090.69    5090.69    5090.69      5090.69   
              75%      6201.25    6160.00    6300.00      6068.75   
              max      9998.00    9998.00    9999.00      9995.00   

Region               North America  South America  
Year          count        5223.00        5245.00  
              mean         2016.99        2017.03  
              std             3.46           3.48  
              min          2010.00        2010.00  
              25%          2015.00        2015.00  
              50%          2017.00        2017.00  
              75%          2019.00        2019.00  
              max          2024.00        2024.00  
Engine_Size_L count        5223.00        5245.00  
              mean            3.24           3.25  
              std             0.81           0.80  
              min             1.50           1.50  
              25%             2.80           2.90  
              50%             3.25           3.25  
              75%             3.60           3.60  
              max             5.00           5.00  
Mileage_KM    count        5223.00        5245.00  
              mean       101310.59      100649.00  
              std         46450.60       46296.06  
              min            92.00          63.00  
              25%         79519.00       78589.00  
              50%        1

In [39]:
# Азия - ДОМИНИРУЮЩИЙ рынок (23,467 записей, 47% данных)
# Остальные регионы равномерны (~5,300 каждый, 10-11%)
# Европа - неожиданно малая выборка (только 5,301)
# НЕТ значимых региональных различий по годам/ 
# Африка и Европа: предпочитают БОЛЬШИЕ двигатели (Q3 = 3.7L)
# Азия и Северная Америка: МЕНЬШИЕ двигатели (Q3 = 3.6L)
# Средний объем: 3.25L везде одинаков
# Африка: МАКСИМАЛЬНЫЙ средний пробег (101,488 км)
# Северная Америка: высокий пробег (101,311 км)  
# Азия: МИНИМАЛЬНЫЙ средний пробег (100,253 км)
# Европа: близко к общему среднему
# Европа: САМЫЕ ДОРОГИЕ авто (среднее $75,404)
# Ближний Восток: вторые по цене ($75,178)
# Африка: САМЫЕ ДЕШЕВЫЕ ($74,884)
# Разница менее 1% - ЦЕНЫ УНИФИЦИРОВАНЫ
# Итог: BMW демонстрирует высокую степень глобальной стандартизации при незначительных региональных особенностях, с явным доминированием азиатского рынка.

In [40]:
# Группирую по региону и виду топлива, считаем среднюю цену и объем продаж и пробег
grouped_region_year = df2.groupby(['Region', 'Fuel_Type']).agg({
    'Price_USD': 'mean',
    'Sales_Volume': 'sum',
    'Mileage_KM': 'mean'
}).round(2)

print("Средние показатели по регионам и топливу:")
print(grouped_region_year.head(51))

Средние показатели по регионам и топливу:
                         Price_USD  Sales_Volume  Mileage_KM
Region        Fuel_Type                                     
Africa        Diesel      75328.12    4230274.54    99910.99
              Electric    75489.33    4256752.69   101927.04
              Hybrid      74589.62   14139674.24   102864.60
              Petrol      74815.06    4383714.79    98219.79
Asia          Diesel      74536.29   18850468.50   101487.57
              Electric    75317.16   18678802.56   101113.74
              Hybrid      75270.84   62753141.13    99636.17
              Petrol      74625.72   19047263.15   100208.03
Europe        Diesel      75122.42    4293009.78    99469.74
              Electric    75686.04    4352090.62   100134.76
              Hybrid      75452.34   14138482.08   100764.04
              Petrol      75232.22    4298284.79   101581.00
Middle East   Diesel      75878.45    4466260.40   101718.43
              Electric    74511.14    44774

In [41]:
# Разница средних цен между регионами и типами топлива менее 2%
#Ближний Восток: самые высокие цены на дизель ($75,878)
#Северная Америка: самые низкие цены на бензин ($74,425)
#Объемы продаж в Азии в 4-5 раз выше, чем в других регионах


In [42]:
# Подсчет электрических автомобилей в Европе
electric_in_europe = df2[(df2['Region'] == 'Europe') & (df2['Fuel_Type'] == 'Electric')].shape[0]
print(f"Электрических автомобилей в Европе: {electric_in_europe}")

Электрических автомобилей в Европе: 863


In [43]:
# Группировка по модели и региону, средняя цена
df_gr1 = round(df.groupby(['Model', 'Region'])['Price_USD'].mean(), 0)
print("Средняя цена по моделям и регионам:")
print(df_gr1.head(15))

Средняя цена по моделям и регионам:
Model     Region       
3 Series  Africa           78009.0
          Asia             75411.0
          Europe           75618.0
          Middle East      73628.0
          North America    77606.0
          South America    73280.0
5 Series  Africa           77431.0
          Asia             73275.0
          Europe           78039.0
          Middle East      73258.0
          North America    76890.0
          South America    73872.0
7 Series  Africa           75312.0
          Asia             77343.0
          Europe           76564.0
Name: Price_USD, dtype: float64


In [44]:
#Ценовая стратегия BMW варьируется по регионам для разных моделей. 3 Series наиболее дорогая в Африке, 
#5 Series - в Европе, что отражает региональные потребительские предпочтения и конкурентную среду.

In [45]:
df_gr_agg_color = df.groupby('Color').agg({
    'Price_USD': ['mean', 'median'],
    'Sales_Volume': ['sum', 'mean'],
    'Mileage_KM': ['mean', 'median'],
    'Year': 'mean'
}).round(2)

print("Анализ по цветам:")
print(df_gr_agg_color)

Анализ по цветам:
       Price_USD          Sales_Volume          Mileage_KM               Year
            mean   median          sum     mean       mean    median     mean
Color                                                                        
Black   75368.82  76379.0   17152682.0  5049.36  100696.40  100958.5  2016.99
Blue    74962.18  74271.5   17372589.0  5126.17  100584.17  100718.5  2017.13
Grey    74393.98  73988.0   17468853.0  5069.31  101322.39  102996.5  2016.95
Red     74629.16  73755.0   17881397.0  5091.51  100408.93  101047.0  2016.94
Silver  75086.22  75547.5   17502439.0  5140.22  100493.99  101412.5  2017.00
White   75470.02  75555.5   16777157.0  5077.83  100473.16  100139.0  2017.03


In [46]:
#Красный цвет BMW - самый популярный по продажам, но белый и черный - самые премиальные по цене. 
#Серый цвет предлагает лучшую стоимость для покупателей. Все цвета имеют практически идентичные технические характеристики, 
#что подтверждает стандартизацию производства BMW независимо от цвета кузова.

# Вывод итоговый: 
КЛЮЧЕВЫЕ ИНСАЙТЫ
1. ГЕОГРАФИЧЕСКОЕ ДОМИНИРОВАНИЕ
Азия является ключевым рынком с 47% всех продаж

Остальные регионы распределены равномерно (~10% каждый)

Европа как домашний рынок представлена неожиданно мало

2. ЦЕНОВАЯ СТРАТЕГИЯ
Высокая стандартизация цен по всему миру (различия менее 1%)

Типичная цена BMW: $65,000-$85,000

Белый и черный цвета самые дорогие, серый - самый доступный

Электрические автомобили не имеют ценового премиума

3. ПОТРЕБИТЕЛЬСКИЕ ПРЕДПОЧТЕНИЯ
Красный цвет - лидер по продажам (17.88 млн единиц)

Азия предпочитает гибридные технологии (62.7 млн продаж)

Африка и Европа выбирают автомобили с большими двигателями

Автоматические трансмиссии доминируют на всех рынках

4. ТЕХНИЧЕСКИЕ ХАРАКТЕРИСТИКИ
Типичный BMW: 2017 год, 3.25L двигатель, 100,591 км пробега

Все цвета имеют идентичные технические характеристики

Пробег обратно коррелирует с ценой автомобиля

5. РЕГИОНАЛЬНЫЕ ОСОБЕННОСТИ
Африка: максимальный пробег + большие двигатели

Европа: самые высокие цены на 5 Series

Северная Америка: максимальные объемы продаж

Ближний Восток: высокий спрос на электрические авто